api/client/trust.go

```diff
@@ -0,0 +1,525 @@
+
+	go func() {
+		targets := []target{}
+		scanner := bufio.NewScanner(r)
+		scanner.Split(ScanInput)
+		// New splitter
+		for scanner.Scan() {
+			line := scanner.Text()
+			if matches := targetRegexp.FindStringSubmatch(line); len(matches) == 4 {
+				dgst, err := digest.ParseDigest(matches[2])
+				if err != nil {
+					fmt.Fprintf(in, "Bad digest value %q, ignoring\n", matches[2])
+					continue
+				}
+				s, err := strconv.ParseInt(matches[3], 10, 64)
+				if err != nil {
+					fmt.Fprintf(in, "Bad size value %q, ignoring\n", matches[3])
+					continue
+				}
+
+				targets = append(targets, target{
+					reference: registry.ParseReference(matches[1]),
+					digest:    dgst,
+					size:      s,
+				})
+			}
+		}
+		targetChan <- targets
+	}()
+
+	type streamWriteCloser struct {
+		io.Writer
+		io.Closer
+	}
+	return streamWriteCloser{
+		Writer: out,
+		Closer: w,
+	}, targetChan
+}
+
+func (cli *DockerCli) trustedPush(repoInfo *registry.RepositoryInfo, tag string, authConfig cliconfig.AuthConfig) error {
+	streamOut, targetChan := targetStream(cli.out)
+
+	v := url.Values{}
+	v.Set("tag", tag)
+
+	_, _, err := cli.clientRequestAttemptLogin("POST", "/images/"+repoInfo.RemoteName+"/push?"+v.Encode(), nil, streamOut, repoInfo.Index, "push")
+	if err != nil {
+		return err
+	}
+
+	// Close stream channel to finish target parsing
+	if err := streamOut.Close(); err != nil {
+		return err
+	}
+
+	// Get target results
+	targets := <-targetChan
+
+	if tag == "" {
+		fmt.Fprintf(cli.out, "No tag specified, skipping TUF metadata push\n")
+		return nil
+	}
+	if len(targets) == 0 {
+		fmt.Fprintf(cli.out, "No targets found, skipping TUF metadata push\n")
+		return nil
+	}
+
+	fmt.Fprintf(cli.out, "Signing and pushing TUF metadata\n")
+
+	repo, err := cli.getNotaryRepository(repoInfo, authConfig)
+	if err != nil {
+		fmt.Fprintf(cli.out, "Error establishing connection to notary repository: %s\n", err)
+		return err
+	}
+
+	for _, target := range targets {
+		hex, err := hex.DecodeString(target.digest.Hex())
+		if err != nil {
+			return err
+		}
+		t := &client.Target{
+			Name: target.reference.String(),
+			Hashes: data.Hashes{
+				string(target.digest.Algorithm()): hex,
+			},
+			Length: int64(target.size),
+		}
+		if err := repo.AddTarget(t); err != nil {
+			return err
+		}
+	}
+
+	passFunc := func() (string, error) {
+		passphrase, err := cli.getPassphrase()
+		if err != nil {
+			return "", err
+		}
+		return passphrase, nil
+	}
+
+	err = repo.Publish(passFunc)
+	if _, ok := err.(*client.ErrRepoNotInitialized); ok {
+		ks, err := keystoremanager.NewKeyStoreManager(cli.trustDirectory())
+		if err != nil {
+			return err
+		}
+		keys := ks.RootKeyStore().ListKeys()
+		var rootKey string
+
+		if len(keys) == 0 {
+			fmt.Fprintf(cli.out, "Creating a new private key\n")
+		} else {
+			// TODO(dmcgowan): let user choose
+			rootKey = keys[0]
+			fmt.Fprintf(cli.out, "Using private key %q\n", rootKey)
+		}
+		passphrase, err := passFunc()
```

Comment:
The requirement I have heard has been:

Only ask for the password when it's required (because it's required infrequently so asking for it every time is annoying).

To achieve that requirement there are 2 options:

1. Make the caller of of the NotaryClient functions more aware of what's going on. - This has been deemed undesirable because it will leak more Notary/TUF related things up the call stack.
2. Enable NotaryClient to retrieve the password if and when it needs it.

With that in mind, what is the most basic thing we need? It's a function that takes no arguments, and returns a password (and maybe an error). That function could be defined on the CryptoService, it could be defined on some new Authenticator struct (an idea Stephen raised last week), but however it is defined, that specific function can still be passed in where it is required.

This implementation is going to be the most stable way to handle this while still leaving open a plethora of options for improving exactly how and when a password is retrieved.